examples/sealed-auctions/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+#![feature(proc_macro_hygiene)]
+
+pub mod types;
+
+extern crate chrono;
+
+use chrono::Utc;
+use std::collections::{HashMap, HashSet}; // `use` statements can go anywhere
+
+#[mantle::service]
+mod service {
+    use crate::types::*;
+
+    #[derive(Service)]
+    pub struct AuctionMarket {
+        /// The item counter
+        itemCounter: usize,
+
+        /// The administrators of this auction mechanism
+        /// Administrators can blacklist users without due cause
+        admins: HashSet<UserId>,
+
+        /// The blacklist of users
+        blacklist: HashSet<UserId>,
+
+        /// The current valuation of the market as the cummulative value
+        /// of all artifacts that have been up for auction
+        marketSize: U256,
+
+        /// The repository of unique artifacts
+        artifacts: HashMap<ItemId, Artifact>,
+
+        /// Active auctions
+        auctions: HashMap<ItemId, Auction>,
+    }
+
+    impl AuctionMarket {
+        /// Constructs a new `AuctionMarket`.
+        pub fn new(ctx: &Context) -> Result<Self> {
+            let mut admins = HashSet::new();
+            admins.insert(ctx.sender());
+
+            let artifacts = HashMap::new();
+            let blacklist = HashSet::new();
+            let auctions = HashMap::new();
+            let itemCounter: usize = 0;
+            let marketSize: U256 = U256::from(0);
+
+            Ok(Self {
+                itemCounter,
+                admins,
+                blacklist,
+                marketSize,
+                artifacts,
+                auctions,
+            })
+        }
+
+        /// Adds a new auction item to the marketplace
+        /// Anyone can sell an item. Once they add an item to the market
+        /// they control the auction through its conclusion
+        pub fn add_item(
+            &mut self,
+            ctx: &Context,
+            value: usize,
+            description: String,
+        ) -> Result<Artifact> {
+            if self.blacklist.contains(&ctx.sender()) {
+                return Err(failure::format_err!(
+                    "The marketplace is not open to {}.",
+                    ctx.sender()
+                ));
+            }
+
+            // build a new artifact object
+            let artifact = Artifact {
+                base: ArtifactBase {
+                    item_id: self.itemCounter,
+                    description: description,
+                },
+                owner: ctx.sender(),
+                provenance: Vec::new(),
+                value: value,
+                transaction_time: Utc::now().to_rfc2822(),
+            };
+            self.artifacts.insert(self.itemCounter, artifact.clone());
+
+            // emit an event to notify watchers that a new item has been added and will
+            // be available for auction soon
+            let base = &self.artifacts.get(&self.itemCounter).unwrap().base;
+            Event::emit(base);
```

Comment:
```rust
Event::emit(&self.artifacts.get(&self.itemCounter).unwrap().base);
```
or
```rust
self.artifacts.get(&self.itemCounter).as_ref().unwrap().base.emit(;
```

Reply:
Picked the former.. it was that way before I thought I would return base and then changed my mind.